Add AI functions (#24963)
trinodb/trino: docs/src/main/sphinx/functions/ai.md

# AI functions

## Overview

The AI functions allow you to invoke a large language model (LLM)
to perform various textual tasks.

> Multiple LLM providers are supported: OpenAI and Ollama.

Reviewer: ```suggestion
Multiple LLM providers are supported, specifically [OpenAI API](https://platform.openai.com/docs/overview) directly and many others like Llama, DeepSeek, Phi, Mistral, Gemma, and others using [Ollama](https://ollama.com/).

The LLM must be provided outside of Trino as an external service.
```

> ## Configuration

Reviewer: I had a look at the code now .. there are a bunch of config properties.. we should add a table with all of them ..

Reviewer: The current content is sufficient to get started. We can do a follow up PR for more docs and examples

Because the AI functions require an external LLM service, they are not

> available by default. To enable them, you must configure them using a

Reviewer: ```suggestion
available by default. To enable them, you must configure a
```

> [catalog properties file](catalog-properties), which will register the functions

Reviewer: ```suggestion
[catalog properties file](catalog-properties) to register the functions invoking the specific LLM
```

> under the specified catalog name.

Reviewer: You can also configure multiple catalogs to use the same functions with different LLM providers.

> Create a catalog properties file `etc/catalog/ai.properties` that references

Reviewer: ```suggestion
Create a catalog properties file `etc/catalog/llm.properties` that references
```

the `ai` connector:

```properties
connector.name=ai
```

> The AI functions all use the `ai` schema name. To avoid needing to reference

Reviewer: ```suggestion
The AI functions all use the `ai` schema name. For the preceding example catalog all functions are available at `llm.ai`.

To avoid needing to reference
```

the functions using their fully qualified name, you can configure the
`sql.path` [SQL environment property](/admin/properties-sql-environment)

> in the `config.properties` file to include the `ai` catalog and schema:

Reviewer: ```suggestion
in the `config.properties` file to include the catalog and schema:
```

```properties

> sql.path=ai.ai

Reviewer: ```suggestion
sql.path=llm.ai
```

```

## Providers

> 

Reviewer: The AI functions invoke an external LLM, which must be configured in the catalog. Performance, results, and cost of all AI function invocations depend on this LLM provider. You must configure a model that is suitable for textual analysis.

### OpenAI

The OpenAI provider uses the [OpenAI API](https://platform.openai.com/docs/overview)
to perform the AI functions:

```properties
ai.provider=openai
ai.model=gpt-4o-mini
ai.openai.api-key=xxx
```

### Ollama

The Ollama provider uses [Ollama](https://ollama.com/) to perform the AI functions:

```properties
ai.provider=ollama
ai.model=llama3.3

> ai.ollama.endpoint=http://localhost:11434

Reviewer: any authentication support @electrum ? Also .. 

```suggestion
ai.ollama.endpoint=http://llm.example.com:11434
```

Since localhost wont work on a cluster unless you run it on each worker.. which we probably dont want

Author: Ollama doesn't support authentication

Author: If users need some sort of authentication (and can't do this at the network level), one option is to use TLS client certificates with `ai.http-client.key-store-path` where Ollama is hosted behind a TLS proxy.

But let's see what people need here. LiteLLM seems to support authentication, so that's probably a good option.

```

## Functions

> 

Reviewer: The following functions are available in each catalog configured with the connector and use the specific LLM.

> :::{function} ai_analyze_sentiment(text) -> varchar

Reviewer 2: Pretty nice showcase of the functionality. I appreciate the effort of documenting it. This will likely get traction for this functionality.

Analyzes the sentiment of the input text.

> The sentiment will be one of

Reviewer: ```suggestion
The sentiment result is 
```

`'positive'`, `'negative'`, `'neutral'`, or `'mixed'`.

```sql
SELECT ai_analyze_sentiment('I love Trino');
-- positive
```
:::

:::{function} ai_classify(text, labels) -> varchar
Classifies the input text according to the provided labels.

```sql
SELECT ai_classify('Buy now!', ARRAY['spam', 'not spam']);
-- spam
```
:::

:::{function} ai_extract(text, labels) -> map(varchar, varchar)
Extracts values for the provided labels from the input text.

```sql
SELECT ai_extract('John is 25 years old', ARRAY['name', 'age']);
-- {name=John, age=25}
```
:::

:::{function} ai_fix_grammar(text) -> varchar
Corrects grammatical errors in the input text.

```sql
SELECT ai_fix_grammar('I are happy. What you doing?');
-- I am happy. What are you doing?
```
:::

> :::{function} ai_gen(prompt) -> varchar

Reviewer: Hm .. might be nice to have this be called ai_generate

Author: Yep, but that's what Databricks named it

Reviewer: Maybe we can add a shim that calls the other for compatibility

Generates text based on the input prompt.

```sql
SELECT ai_gen('Describe Trino in a few words');
-- Distributed SQL query engine.
```
:::

:::{function} ai_mask(text, labels) -> varchar
Masks values for the provided labels in the input text.

```sql
SELECT ai_mask(
    'Contact me at 555-1234 or visit us at 123 Main St.',
    ARRAY['phone', 'address']);

> -- Contact me at [MASKED] or visit us at [MASKED].

Reviewer: Does it literally use this `[MASKED]` string .. should this be configurable or potentially another optional parameter?

Author: It is literally `[MASKED]`, per the Databricks definition

Reviewer: We can do a follow up to make it configurable later

different LLM providers. In this case, the functions will need to be
referenced in queries using their fully qualified name, rather than
relying on the SQL path.

## Providers

The AI functions invoke an external LLM, which must be configured for the
catalog. Performance, results, and cost of all AI function invocations are
completely dependent on the LLM provider and the model used.
You must specify a model that is suitable for textual analysis.

### Anthropic

The Anthropic provider uses the [Anthropic API](https://www.anthropic.com/api)
to perform the AI functions:

```properties
ai.provider=anthropic
ai.model=claude-3-5-sonnet-latest
ai.anthropic.api-key=xxx
```

### OpenAI

The OpenAI provider uses the [OpenAI API](https://platform.openai.com/)
to perform the AI functions:

```properties
ai.provider=openai
ai.model=gpt-4o-mini
ai.openai.api-key=xxx
```

### Ollama

The OpenAI provider can be used with [Ollama](https://ollama.com/)
to perform the AI functions, as Ollama is compatible with the OpenAI API:

```properties
ai.provider=openai
ai.model=llama3.3
ai.openai.endpoint=http://localhost:11434
ai.openai.api-key=none
```

> The API key must be specified, but will be ignored by Ollama.

Author: There is no need for it to be random. That would be misleading, as it would seem to be a valid key.

> The API key must be specified, but will be ignored by Ollama.

Reviewer: Yeah .. its misleading in that sense .. sorry .. I mean like you used "none" .. it could "any:, "foo" , or any other random string..

> The API key must be specified, but will be ignored by Ollama.

Author: I think we're over thinking this 😄

> The API key must be specified, but will be ignored by Ollama.

Reviewer: Hahah .. yeah.
